File: AtCoder/ABC/170/python/D_Not_Divisible.py
MAX_A = int(1e6)


N = int(input())
A = list(map(int, input().split()))


def divisors(n):
    res = []
    for i in range(1, int(n ** 0.5) + 1):
        if n % i == 0:
            res.append(i)
            if n // i != i:
                res.append(n // i)
    return res


cnt = [0] * (MAX_A + 1)
for a in A:
    cnt[a] += 1
res = 0
for a in A:
    v = divisors(a)
    found = False
    for d in v:
        if a == d and cnt[a] == 1:
            continue
        if cnt[d] > 0:
            found = True
            break
    if not found:
        res += 1
print(res)

File: AtCoder/ABC/170/cpp/D_Not_Divisible.cpp
#include <bits/stdc++.h>
#define MAX_N 200'000
#define MAX_A 1'000'000
using namespace std;

int N;
int A[MAX_N];
int cnt[MAX_A + 1];

vector<int> divisors(int n) {
  vector<int> res;
  for (int i = 1; i * i <= n; ++i) {
    if (n % i == 0) {
      res.push_back(i);
      if (n / i != i) {
        res.push_back(n / i);
      }
    }
  }
  return res;
}

void solve() {
  for (int i = 0; i < N; ++i) {
    ++cnt[A[i]];
  }
  int res = 0;
  for (int i = 0; i < N; ++i) {
    vector<int> v = divisors(A[i]);
    bool found = false;
    for (auto n : v) {
      if (n == A[i] && cnt[n] == 1) {
        continue;
      }
      if (cnt[n] > 0) {
        found = true;
        break;
      }
    }
    if (!found) {
      ++res;
    }
  }
  printf("%d\n", res);
}

int main() {
  cin >> N;
  for (int i = 0; i < N; ++i) {
    cin >> A[i];
  }
  solve();
}